### PR TITLE
LibJS: Add several Temporal.Instant methods

### PR DESCRIPTION
**Pull request by IdanHo**
This fixes 0 test262 test cases :^(
(Sorry for the large commits :sweat_smile:)

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.h [hunk 1]
```diff
@@ -20,6 +20,14 @@ enum class OptionType {
 };
 Value get_option(GlobalObject&, Object& options, String const& property, Vector<OptionType> const& types, Vector<StringView> const& values, Value fallback);
 
+String to_temporal_rounding_mode(GlobalObject&, Object& normalized_options, String const& fallback);
+
+u64 to_temporal_rounding_increment(GlobalObject&, Object& normalized_options, Optional<double> dividend, bool inclusive);
+
+Optional<String> to_smallest_temporal_unit(GlobalObject&, Object& normalized_options, Vector<StringView> const& disallowed_units, Optional<String> fallback);
+
+BigInt* round_number_to_increment(GlobalObject&, BigInt const&, u64 increment, String const& rounding_mode);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.h:29]: Don't think these empty lines are necessary, would just mush them together :)

File: Userland/Libraries/LibJS/Runtime/Temporal/InstantPrototype.cpp [hunk 2]
```diff
@@ -122,6 +124,87 @@ JS_DEFINE_NATIVE_FUNCTION(InstantPrototype::epoch_nanoseconds_getter)
...
+    }
+    // 11. Else if smallestUnit is "microsecond", then
+    else if (smallest_unit == "microsecond"sv) {
+        // a. Let maximum be 8.64 × 10^10.
+        maximum = 86400000000;
+    }
+    // 12. Else,
+    else {
+        // a. Assert: smallestUnit is "nanosecond".
+        dbgln("{}", smallest_unit);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/InstantPrototype.cpp:188]: Leftover from debugging?

**Review by linusg — COMMENTED** [review]
Fantastic work!

---

Typo in second commit: `it's` => `its`

File: Userland/Libraries/LibJS/Runtime/Date.cpp [hunk 3]
```diff
@@ -111,4 +111,86 @@ String Date::iso_date_string() const
     return builder.build();
 }
 
+// https://tc39.es/ecma262/#eqn-msPerSecond
+static constexpr double ms_per_second = 1000;
+// https://tc39.es/ecma262/#eqn-msPerMinute
+static constexpr double ms_per_minute = 1000;
+// https://tc39.es/ecma262/#eqn-msPerHour
+static constexpr double ms_per_hour = 1000;
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibJS/Runtime/Date.cpp:119]: those values don't look quite right

File: Userland/Libraries/LibJS/Runtime/Temporal/Instant.cpp [hunk 4]
```diff
@@ -65,4 +68,73 @@ Object* create_temporal_instant(GlobalObject& global_object, BigInt& epoch_nanos
...
+    // 1. If Type(item) is Object, then
+    if (item.is_object()) {
+        // a. If item has an [[InitializedTemporalInstant]] internal slot, then
+        if (is<Instant>(item.as_object())) {
+            // i. Return item.
+            return &static_cast<Instant&>(item.as_object());
+        }
+        // TODO:
+        // b. If item has an [[InitializedTemporalZonedDateTime]] internal slot, then
+        // i. Return ! CreateTemporalInstant(item.[[Nanoseconds]]).
```
[the remark this answers is not in the document]
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/Temporal/Instant.cpp:85]: The issue is that the TODO is for the if, aka we can't detect an object with a "InitializedTemporalZonedDateTime", so we can't conditionally do a TODO/return undefined in that case either.

[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/Temporal/Instant.cpp:85]: Yeah, but we should definitely return from the is_object branch

[the remark this answers is not in the document]
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/Temporal/Instant.cpp:85]: Not necessarily? The spec allows it to be coerced into a string and then parsed.

[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/Temporal/Instant.cpp:85]: oh, I see now... you're right, nevermind 😅

File: Userland/Libraries/LibJS/Runtime/Date.cpp (hunk 3 above)
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/Date.cpp:119]: oops, fixed!

File: Userland/Libraries/LibJS/Runtime/Temporal/InstantPrototype.cpp (hunk 2 above)
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/Temporal/InstantPrototype.cpp:188]: yep, my bad, removed

File: Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.h (hunk 1 above)
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.h:29]: done :)

File: Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.cpp [hunk 5]
```diff
@@ -13,6 +13,105 @@
...
+
+    // 7. If type is Boolean, then
+    if (type == OptionType::Boolean) {
+        // a. Set value to ! ToBoolean(value).
+        value = Value(value.to_boolean());
+    }
+    // 8. Else if type is Number, then
+    else if (type == OptionType::Number) {
+        // a. Set value to ? ToNumber(value).
+        value = value.to_number(global_object);
```
[the remark this answers is not in the document]
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.cpp:86]: my bad, fixed.

File: Userland/Libraries/LibJS/Runtime/Temporal/InstantConstructor.cpp [hunk 6]
```diff
@@ -68,6 +69,19 @@ Value InstantConstructor::construct(FunctionObject& new_target)
...
 }
 
+// 8.2.2 Temporal.Instant.from ( item ), https://tc39.es/proposal-temporal/#sec-temporal.instant.from
+JS_DEFINE_NATIVE_FUNCTION(InstantConstructor::from)
+{
+    auto item = vm.argument(0);
+
+    // 1. If Type(item) is Object and item has an [[InitializedTemporalInstant]] internal slot, then
+    if (item.is_object() && is<Instant>(item.as_object()))
+        return create_temporal_instant(global_object, *js_bigint(vm.heap(), static_cast<Instant&>(item.as_object()).nanoseconds().big_integer()));
```
[the remark this answers is not in the document]
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/Temporal/InstantConstructor.cpp:79]: yep, sorry, added.

File: Userland/Libraries/LibJS/Runtime/Temporal/Instant.cpp [hunk 7]
```diff
@@ -137,4 +137,19 @@ BigInt* parse_temporal_instant(GlobalObject& global_object, String const& iso_st
     return js_bigint(vm.heap(), utc->big_integer().minus(Crypto::SignedBigInteger::create_from(offset_nanoseconds)));
 }
 
+// 8.5.5 CompareEpochNanoseconds ( epochNanosecondsOne, epochNanosecondsTwo )
+i32 compare_epoch_nanoseconds(BigInt const& one, BigInt const& two)
```
[the remark this answers is not in the document]
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/Temporal/Instant.cpp:141]: yep, done

File: Userland/Libraries/LibJS/Runtime/Temporal/TimeZone.cpp [hunk 8]
```diff
@@ -92,4 +93,145 @@ Object* create_temporal_time_zone(GlobalObject& global_object, String const& ide
     return object;
 }
 
+// https://tc39.es/proposal-temporal/#prod-TimeZoneNumericUTCOffset
```
[the remark this answers is not in the document]
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/Temporal/TimeZone.cpp:96]: yep, done, we might want to split these off to `Syntax.cpp` or something in the future once we add parsing for all of the other temporal related syntax.

File: Userland/Libraries/LibJS/Runtime/Date.cpp [hunk 9]
```diff
@@ -111,4 +111,86 @@ String Date::iso_date_string() const
...
+// https://tc39.es/ecma262/#eqn-msPerDay
+static constexpr double ms_per_day = 86400000;
```
[the remark this answers is not in the document]
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/Date.cpp:121]: done!

File: Userland/Libraries/LibJS/Runtime/Temporal/Instant.cpp [hunk 10]
```diff
@@ -7,8 +7,11 @@
 #include <LibCrypto/BigInt/SignedBigInteger.h>
 #include <LibJS/Runtime/AbstractOperations.h>
 #include <LibJS/Runtime/GlobalObject.h>
+#include <LibJS/Runtime/Temporal/AbstractOperations.h>
```
[the remark this answers is not in the document]
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/Temporal/Instant.cpp:10]: if you insist :^)

File: AK/DateTimeLexer.h [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+namespace AK {
+
+class DateTimeLexer : public GenericLexer {
+public:
+    constexpr explicit DateTimeLexer(const StringView& input)
+        : GenericLexer(input)
+    {
+    }
+
+    StringView consume_year()
```
[the remark this answers is not in the document]
Comment (IdanHo) [reply AK/DateTimeLexer.h:21]: done! :)

**Pull request merged**